core/loop_mixin.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview The mixin that holds logic for setting a warning on a block if
+ * it is not inside a loop.
+ */
+'use strict';
+
+goog.module('Blockly.loopMixin');
+
+/* eslint-disable-next-line no-unused-vars */
+const Abstract = goog.requireType('Blockly.Events.Abstract');
+const Events = goog.require('Blockly.Events');
+const Extensions = goog.require('Blockly.Extensions');
+const Msg = goog.require('Blockly.Msg');
+/* eslint-disable-next-line no-unused-vars */
+const {Block} = goog.requireType('Blockly.Block');
+
+
+/**
+ * This mixin adds a check to make sure the 'controls_flow_statements' block
+ * is contained in a loop. Otherwise a warning is added to the block.
+ * @mixin
+ * @augments Block
+ * @readonly
+ */
+const CONTROL_FLOW_IN_LOOP_CHECK_MIXIN = {
+  /**
+   * List of block types that are loops and thus do not need warnings.
+   * To add a new loop type add this to your code:
+   * Blockly.loopMixin.CONTROL_FLOW_IN_LOOP_CHECK_MIXIN.LOOP_TYPES.push('custom_loop');
+   *
+   * Please leave a comment on this github issue
+   * https://github.com/mit-cml/appinventor-sources/issues/2032 if you end up
+   * needing to use this.
+   */
+  LOOP_TYPES: [
+    'controls_repeat',
+    'controls_repeat_ext',
+    'controls_forEach',
+    'controls_for',
+    'controls_whileUntil',
```

Comment:
I admit that I probably did not fully grok our recent discussion about this change—it was getting rather late here—but looking at this resulting PR I am completely perplexed about why this code has been moved to `core/`, since, for example, these block types are totally meaningless without the definitions given in `blocks/`.

Since modules in `generators/` require modules in `blocks/`, and since it would be easy to enable external code to `import {CONTROL_FLOW_IN_LOOP_CHECK_MIXIN} from 'blockly/blocks';` I don't really know why it wouldn't be better to leave this code in `blocks/loops.js` (but move it to the `Blockly.blocks.loops` module, and reexport from `Blockly.loops.all`)

Reply:
The idea that some blocks can contain break or continue, and others can't, seemed like a core idea that had gotten pushed out into the blocks. Although perhaps the controls block definition file should push each of these names onto the list as it defines the block, rather than having it hardcoded from the start.